substrate-node/pallets/pallet-smart-contract/src/tests.rs

```diff
@@ -3123,6 +3123,103 @@ fn test_percent() {
     assert_eq!(new_cost, 248);
 }
 
+#[test]
+fn test_calculate_discount_works() {
+    let amount_due = 1000;
+    let seconds_elapsed = SECS_PER_HOUR; // amount due is relative to 1h
+    let balance = amount_due * 24 * 30; // just enough balance for 1 month at the rate of 1000 per hour
+
+    // No discount level
+    let result = cost::calculate_discount::<TestRuntime>(
+        amount_due,
+        seconds_elapsed,
+        balance,
+        NodeCertification::Diy,
+    );
+    let expected_discount_level = types::DiscountLevel::None;
+    assert_eq!(
+        result,
+        (
+            (U64F64::from_num(amount_due) * expected_discount_level.price_multiplier())
+                .ceil()
+                .to_num::<u64>(),
+            expected_discount_level
+        )
+    );
+
+    // Default discount level
+    let result = cost::calculate_discount::<TestRuntime>(
+        amount_due,
+        seconds_elapsed,
+        balance * 3, // just enough balance for 3 month
+        NodeCertification::Diy,
+    );
+    let expected_discount_level = types::DiscountLevel::Default;
+    assert_eq!(
+        result,
+        (
+            (U64F64::from_num(amount_due) * expected_discount_level.price_multiplier())
+                .ceil()
+                .to_num::<u64>(),
+            expected_discount_level
+        )
+    );
+
+    // Bronze discount level
+    let result = cost::calculate_discount::<TestRuntime>(
```

Comment:
Same here and the ones below too. Each test should test one specific case. You can usually split those tests up into 3 pieces: prepare, execute, assert. If you don't have that you're probably testing too much in one test. The reason why this is so important is that it is very easy to navigate through/understand tests like that. Another solution here would be parameterized tests. Rust doesn't support that by default but you can use macros though: https://stackoverflow.com/questions/34662713/how-can-i-create-parameterized-tests-in-rust. Parameterized tests are good if you want to avoid dupplication of test code like in this example.

Reply:
Thx for the parameterized tests suggestion
More elegant now
See below